src/sagemaker/utils.py

```diff
@@ -1494,23 +1493,89 @@ def format_tags(tags: Tags) -> List[TagsDict]:
+    normpath - normalizes paths (e.g. remove redudant separators)
+    and handles platform-specific differences
+    """
+    return normpath(realpath(abspath(path)))
 
-def check_tarfile_data_filter_attribute():
-    """Check if tarfile has data_filter utility.
 
-    Tarfile-data_filter utility has guardrails against untrusted de-serialisation.
+def _is_bad_path(path, base):
+    """Checks if the joined path (base directory + file path) is rooted under the base directory
 
-    Raises:
-        PythonVersionError: if `tarfile.data_filter` is not available.
+    Ensuring that the file does not attempt to access paths
+    outside the expected directory structure.
+
+    Args:
+        path (str): The file path.
+        base (str): The base directory.
+
+    Returns:
+        bool: True if the path is not rooted under the base directory, False otherwise.
     """
-    # The function and it's usages can be deprecated post support of python >= 3.12
-    if not hasattr(tarfile, "data_filter"):
-        raise PythonVersionError(
-            f"Since tarfile extraction is unsafe the operation is prohibited "
-            f"per PEP-721. Please update your Python [{sys.version}] "
-            f"to latest patch [refer to https://www.python.org/downloads/] "
-            f"to consume the security patch"
-        )
+    # joinpath will ignore base if path is absolute
+    return not _get_resolved_path(joinpath(base, path)).startswith(base)
+
+
+def _is_bad_link(info, base):
+    """Checks if the link is rooted under the base directory.
+
+    Ensuring that the link does not attempt to access paths outside the expected directory structure
+
+    Args:
+        info (tarfile.TarInfo): The tar file info.
+        base (str): The base directory.
+
+    Returns:
+        bool: True if the link is not rooted under the base directory, False otherwise.
+    """
+    # Links are interpreted relative to the directory containing the link
+    tip = _get_resolved_path(joinpath(base, dirname(info.name)))
+    return _is_bad_path(info.linkname, base=tip)
+
+
+def safe_members(members):
+    """A generator that yields members that are safe to extract.
+
+    It checks for bad paths and bad links.
+
+    Args:
+        members (list): A list of members to check.
+
+    Yields:
+        tarfile.TarInfo: The tar file info.
+    """
+    base = _get_resolved_path(".")
+
+    for file_info in members:
+        if _is_bad_path(file_info.name, base):
+            print(stderr, file_info.name, "is blocked (illegal path)")
+        elif file_info.issym() and _is_bad_link(file_info, base):
+            print(stderr, file_info.name, "is blocked: Symlink to", file_info.linkname)
+        elif file_info.islnk() and _is_bad_link(file_info, base):
+            print(stderr, file_info.name, "is blocked: Hard link to", file_info.linkname)
+        else:
+            yield file_info
+
+
+def custom_extractall_tarfile(tar, extract_path):
```

Comment:
Let's also add a todo note that we can deprecate this method once we upgrade SM PySDK to python3.12+

Reply:
done